File: systemProgramming/pythonAssembler/source/src/SymbolTable.py
class SymbolInTable(Exception):
    def __str__(self):
        print("Symbol이 List 내에 이미 존재합니다.")


class SymbolTable:
    symbolList = None
    locationList = None

    def __init__(self):
        self.symbolList = []
        self.locationList = []

    def putSymbol(self, symbol, location):
        if self.search(symbol) != -1:
            raise SymbolInTable
        else:
            self.symbolList.append(symbol)
            self.locationList.append(location)

    def search(self, symbol):
        if symbol in self.symbolList:
            return self.locationList[self.symbolList.index(symbol)]
        else:
            return -1

File: systemProgramming/javaAssembler_sim/source/src/SP20_simulator/SymbolTable.java
package SP20_simulator;
import java.util.ArrayList;


public class SymbolTable {
	ArrayList<String> symbolList = new ArrayList<String>();
	ArrayList<Integer> addressList = new ArrayList<Integer>();
	

	public void putSymbol(String symbol, int address) {
		if(!symbolList.isEmpty()) { //중복확인
			if(search(symbol) != -1)
				modifySymbol(symbol, address);
			else {	
				symbolList.add(symbol);
				addressList.add(address);
			}
		}
		else {
			symbolList.add(symbol);
			addressList.add(address);
		}
	}
	

	public void modifySymbol(String symbol, int newaddress) {
		addressList.set(symbolList.indexOf(symbol), newaddress);
	}
	

	public int search(String symbol) {
		int address = -1;

		for(String p_symbol : symbolList) {
			if(p_symbol.equals(symbol)) {
				address = addressList.get(symbolList.indexOf(p_symbol));
				break;
			}
		}
		return address;
	}
	
	public void clear() {
		symbolList.clear();
		addressList.clear();
	}
	
}
